### DISCUSSION TITLE
Configure renovateBot for check helm release deploy with terraform

### DISCUSSION THREAD

**Discussion by Grounz** [renovatebot/renovate#16052] [Request Help] [answered]
Hi

We want use RenovateBot for check version of helm charts deployed with `helm_release` module in terraform.

below our gitlab terraform code

```shell
____main.tf # Principal main.tf configuration
|____module_argocd
| |____versions.tf
| |____variables.tf # contains terraform vars
| |____README.md
| |____main.tf # Contains helm_release module and version
```
In each module we cal `helm_release` module as below:

```shell
resource "helm_release" "argocd" {
  provider         = helm
  name             = var.argocd_helm_release_name
  repository       = var.argocd_helm_repository
  chart            = var.argocd_helm_chart_name
  version          = var.argocd_helm_version
  namespace        = var.argocd_helm_namespace
  create_namespace = var.argocd_create_namespace
}
```

We have configured renovateBot for check terraform with `depTyes` `helm_release`. But when we check this repository it not found version. 

Below renovate.json file:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    ":semanticCommits",
    "docker:disable"
  ],
  "terraform": {                                                                
    "depTypes": ["helm_release"]                                               
  }   
}
```

So how to configure renovateBot for check `helm_release` deploy by terraform ?

**Comment by viceice** [comment]
you can't use variables. if you use variables, you need to use regex manager to match your variables

**Reply by lenglet-k** [reply]
@viceice  Thanks for you answer. The regex Manager must be configure in renovate.json in git repository ? 

i will try that

**Reply by viceice** [reply]
yes

**Reply by lenglet-k** [reply]
I have try it but i don't understand how to regexManagers work:

Below is my test:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    ":semanticCommits",
    "docker:disable"
  ],
  "docker": {
    "enabled": false
  },
  "terraform": {                                                                
    "depTypes": ["helm_release"]                                               
  },
  "regexManagers": [
    {
      "fileMatch": ["^variables.tf$"],
      "matchStrings": ["default = (?<currentValue>.+)\\n"],
      "depNameTemplate": "terraform",
      "datasourceTemplate": "helm"
    }
  ]
}
```

If i've well understood:

`fileMatch` is regex to find file with my version of dependency, in my case my all helm version of charts are in `variables.tf`, it looks to that:

```
variable "argocd_helm_version" {
  type    = string
  default = "4.8.0"
}
```
`MatchString` is regex to find version [regex101TestRegexp](https://regex101.com/r/Fzhs2C/1)
`depName` is terraform because my check is on terraform file, maybe i'm wrong ?
`datasourceTemplate` i check datasource helm for check new version of helm charts.

Thansk for your time

**Reply by lenglet-k** [reply]
@viceice  Ok i have finally understood. With this configuration is ok

```json
  "regexManagers": [
    {
      "fileMatch": ["\\.tf$"],
      "matchStrings": [
        "default = \"(?<currentValue>([0-9]+)\\.([0-9]+)\\.([0-9]+))\"\\n}\\n\\nvariable \".+_helm_chart_name\" {\\n.+\\n  default = \"(?<depName>[a-z].+)\"\\n}\\n\\n.+\\n.+\\n.+\\n}\\n\\nvariable \".+_helm_repository\" {\\n.+\\n  default = \"(?<registryUrl>.*?)\""
      ],
      "datasourceTemplate": "helm"
    }
```

My variables.tf files contains many information and it's necessary to parse correctly file for that renovate work.

The Workaround: RTFM ;)

**Reply by viceice** [reply]
You should create a new top comment with a full sample and mark as solution for others

**Reply by lenglet-k** [reply]
I create a new discussion for that ? It's first time for me

**Reply by rarkins** [reply]
Create a new top level comment

**Comment by Grounz** [comment, marked as answer]
The solution is below:

For check versions of `Helm` chart deployed by  `terraform`  withf `helm_release` module and when your helm version are set by terraform variables. You should use `regexManagers` of `renovateBot`.

My `main.tf` terraform file:

```tf
module "sentry" {
  source                                   = "./module_sentry"
  sentry_helm_release_name = var.sentry_helm_release_name
  sentry_helm_repository        = var.sentry_helm_repository
  sentry_helm_chart_name     = var.sentry_helm_chart_name
  sentry_helm_version            = var.sentry_helm_version
  sentry_helm_namespace     = var.sentry_helm_namespace
  sentry_create_namespace   = var.sentry_create_namespace
  providers = {
    helm = helm
  }
}
```

Exemple of `variables.tf` in terraform module:

```tf
variable "sentry_helm_version" {
  type    = string
  default = "14.1.2"
}

variable "sentry_helm_chart_name" {
  type    = string
  default = "sentry"
}

variable "sentry_helm_namespace" {
  type    = string
  default = "sentry"
}

variable "sentry_helm_repository" {
  type    = string
  default = "https://sentry-kubernetes.github.io/charts"
}
```

In fact, you should configure a `matchStrings` which capture groups  all elements for `helm` datasources.

Below i capture elements with regexManager for sentry release:

```json
...
  "regexManagers": [
    {
      "fileMatch": ["\\.tf$"],
      "matchStrings": [
        "default = \"(?<currentValue>([0-9]+)\\.([0-9]+)\\.([0-9]+))\"\\n}\\n\\nvariable \".+_helm_chart_name\" {\\n.+\\n  default = \"(?<depName>[a-z].+)\"\\n}\\n\\n.+\\n.+\\n.+\\n}\\n\\nvariable \".+_helm_repository\" {\\n.+\\n  default = \"(?<registryUrl>.*?)\""
      ],
      "datasourceTemplate": "helm"
    }
  ]
```

The first group captured by regex is `currentValue` and it match `default` value of `sentry_helm_version` variable, then the second group is `depName` and it capture `default` value of `sentry_helm_chart_name` and finally the regex capture `registryUrl` value of `sentry_helm_repository` variable.

**Reply by cloudmatt** [reply]
Curious if you have an example of this publicly that I could reference? I have my terraform and renovate exactly like the example shown here but it's not actually working. I don't know if it's a regex issue or what but before I sunk a bunch of time into troubleshooting it was just curious if you had something publicly available that I could use as a reference for.

**Comment by jarihaikonen** [comment]
Made the following to get this working the way I wanted

For variables, you can specify helm releases like this, note that you need the "_helm" part on the variable for the regex to work
```
variable "external_dns_helm" {
  type = object({
    repository = optional(string, "https://charts.bitnami.com/bitnami")
    chart      = optional(string, "external-dns")
    version    = optional(string, "6.26.3")
  })
}
```

and then for the customManagers (current way on renovate)
```
"customManagers": [
      {
        fileMatch: ['\\.tf$'],
        matchStrings: [
          '_helm".+\n.+\n.+repository.+"(?<registryUrl>[a-z].+)\".+\n.+chart.+"(?<depName>[a-z].+)\".+\n.+version.+"(?<currentValue>[^"].+)\"',
        ],
        datasourceTemplate: 'helm',
      },
    ]
```

and you can group those by
```
packageRules: [
      {
        "matchManagers": [
          "regex"
        ],
        "groupName": "Terraform helm packages",
        "groupSlug": "terraform_helm"
      },
     ]
```

**Comment by phyzical** [comment]
The above suggestions work, but for me the .+ messed with things as there may not be spaces, try .* which allows for no spaces if you have trouble 

```
matchStrings: [
          '.*repository.*"(?<registryUrl>[a-z].*)\".*\n.*chart.*"(?<depName>[a-z].*)\".*\n.*version.*"(?<currentValue>[^"].*)\"',
        ],
````
